### PR TITLE
Refactor the implementation of -H option in lsfd

### PR DESCRIPTION
**Pull request by masatake**
Before implementing -H options to the other ls* commands, I revised that in lsfd.

The original code does the following:

```
		fprintf(out, " %20s  %-10s%s\n", infos[i].name,
```

I wanted to remove "20" and "10".
Do you think using libscols as the replacement is overkiling?

### CODE REVIEW

**Comment by t-8ch** [comment]
> I wanted to remove "20" and "10".

If it is only about those two constants something like this would be enough:

```diff
diff --git a/misc-utils/lsfd.c b/misc-utils/lsfd.c
index 4671a2149c55..6e4e7a37fc3f 100644
--- a/misc-utils/lsfd.c
+++ b/misc-utils/lsfd.c
@@ -1794,9 +1794,13 @@ static void collect_processes(struct lsfd_control *ctl, const pid_t pids[], int
 
 static void __attribute__((__noreturn__)) list_colunms(FILE *out)
 {
+       int name_width = 0;
+       for (size_t i = 0; i < ARRAY_SIZE(infos); i++)
+               name_width = max(name_width, (int) strlen(infos[i].name));
+
        fputs(USAGE_COLUMNS, out);
        for (size_t i = 0; i < ARRAY_SIZE(infos); i++)
-               fprintf(out, " %20s  %-10s%s\n", infos[i].name,
+               fprintf(out, " %*s  %-10s%s\n", name_width, infos[i].name,
                        infos[i].json_type == SCOLS_JSON_STRING ?  "<string>":
                        infos[i].json_type == SCOLS_JSON_ARRAY_STRING ?  "<string>":
                        infos[i].json_type == SCOLS_JSON_ARRAY_NUMBER ?  "<string>":
```

**Comment by karelzak** [comment]
I like the idea that the caller can get the columns in JSON format :-) 
```
   lsfd --list-columns --json
```
It could be attractive because applications that consume our ls-like tools do not have to hardcode anything or depend on version string.

File: include/column-list-table.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+#ifndef UTIL_LINUX_COLUMN_LIST_TABLE_H
+#define UTIL_LINUX_COLUMN_LIST_TABLE_H
+
+#include "libsmartcols.h"
+#include <stdio.h>
+
+enum { CLT_COL_HOLDER, CLT_COL_TYPE, CLT_COL_DESC };
+
+static struct libscols_table *xcolumn_list_table_new(const char *table_name,
```
Comment (karelzak) [inline include/column-list-table.h:30]: Maybe "static inline" if we have it in .h file.

File: include/column-list-table.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+{
+	struct clt_colinfo {
+		const char *name;
+		int flags;
+	};
+	const struct clt_colinfo clt_infos[] = {
+		[CLT_COL_HOLDER] = { "HOLDER", SCOLS_FL_RIGHT },
+		[CLT_COL_TYPE]   = { "TYPE",   0 },
+		[CLT_COL_DESC]   = { "DESC",   0 },
+	};
```
Comment (karelzak) [inline include/column-list-table.h:43]: Data-based programming is the best, but I think `struct clt_colinfo` is overkill in this case. All you need is to call scols_table_new_column() three times.
```C
if (!scols_table_new_column(tb, "HOLDER", 0, SCOLS_FL_RIGHT))
  goto failed;
if (!scols_table_new_column(tb, "TYPE", 0, 0))
  goto failed;
if (!scols_table_new_column(tb, "DESCRIPTION", 0, 0))
  goto failed;
```
you do not need scols_column_set_json_type(cl, SCOLS_JSON_STRING); the string is default.

File: include/column-list-table.h [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+	if (!ln)
+		err(EXIT_FAILURE, _("failed to allocate output line"));
+
+	if (scols_line_set_data(ln, CLT_COL_HOLDER, name))
+		err(EXIT_FAILURE, _("failed to add output data"));
+	if (scols_line_set_data(ln, CLT_COL_TYPE, (json_type == SCOLS_JSON_STRING ?        "<string>":
+						   json_type == SCOLS_JSON_ARRAY_STRING ?  "<string>":
+						   json_type == SCOLS_JSON_ARRAY_NUMBER ?  "<string>":
+						   json_type == SCOLS_JSON_NUMBER ?        "<integer>":
+						   json_type == SCOLS_JSON_FLOAT ?         "<float>":
```
Comment (karelzak) [inline include/column-list-table.h:84]: json_type == SCOLS_JSON_BOOLEAN ? <boolean>

File: include/column-list-table.h [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+						   fallback_typename)))
```
Comment (karelzak) [inline include/column-list-table.h:85]:
```
fallback_typename ? : "<string>"
```

**Comment by masatake** [comment]
@t-8ch, @karelzak Thank you for reviewing.

I follow @karelzak's comments.
As a trial, I will apply column-list-table.h to lsblk.

**Comment by masatake** [comment]
The way of setting column types in lsblk is slightly complicated than I expected.

>  As a trial, I will apply column-list-table.h to lsblk.

I withdraw what I wrote.

**Comment by masatake** [comment]
Instead, I'll target lslocks.

**Comment by karelzak** [comment]
Yes, the same lsblk column could be a string and a number; it depends on --bytes or if the column is used in a filter. I think we can resolve this complication in applications (rather than in the header file) by fallback_typename = "<string|number>" and, after the list, adding a note to explain it to the users. I'll do it do lsblk :-)

**Comment by masatake** [comment]
@karelzak I implemented what you wrote in lslocks:

```
[yamato@dev64]~/var/util-linux/util-linux% ./lslocks -H
COMMAND <string>        command of the process holding the lock
    PID <integer>       PID of the process holding the lock
   TYPE <string>        kind of lock
   SIZE <string|number> size of the lock
  INODE <integer>       inode number
MAJ:MIN <string>        major:minor device number
   MODE <string>        lock access mode
      M <boolean>       mandatory state of the lock: 0 (none), 1 (set)
  START <integer>       relative byte offset of the lock
    END <integer>       ending offset of the lock
   PATH <string>        path of the locked file
BLOCKER <integer>       PID of the process blocking the lock
HOLDERS <string>        HOLDERS of the lock
```

File: include/column-list-table.h (hunk 4 above)
Comment (masatake) [reply include/column-list-table.h]: Now I understand how this suggestion is important.

**Pull request merged**